lumicks/pylake/tests/test_channels.py

```diff
@@ -349,15 +349,15 @@ def test_downsampling_to():
     assert np.all(np.equal(s2a.timestamps, [0, 20, 40, 60, 80, 100, 120]))
     assert np.allclose(s2a.data, [2.5, 6.5, 10.0, 12.5, 14.5, 16.5, 18.5])
     s2b = s.downsampled_to(0.5e8, where="center")
-    assert np.all(np.equal(s2b.timestamps, [10, 30, 50, 70, 90, 110, 130]))
+    assert np.all(np.equal(s2b.timestamps, np.array([15/2, 55/2, 90/2, 130/2, 170/2, 210/2, 250/2], dtype=np.int)))
```

Comment:
The behavior here to me seems odd. As a user, I would expect that the `downsampled_to` method would return a continuous slice with timesteps corresponding to the requested frequency. However, in this implementation (and this test specifically) the timesteps are not consistent (20, 17.5, 20, 20, 20, 20). I think for long slices, repeated offsets like this could result in the end of the slice being "out of phase" in a sense with the beginning of the slice

Reply:
I don't really think so. Please correct me if I'm wrong, but while the variable delta timestamp is an annoying inconvenience, if you just use the timestamps as they are computed now, the new way is the more correct one (see figures below). I agree with you that if you assume a constant timestep, where there is none you'd go out of phase, but this doesn't happen if you make sure you use the actual timestamps. This is also why `downsampled_over` always returns a `TimeSeries` and not a `Continuous`.

If we approach the limit of low samples per new sample we get the following:

![Old_ds](https://user-images.githubusercontent.com/19836026/96561609-13caa300-12c0-11eb-9d34-6ef0d7489a93.png)

![New_ds](https://user-images.githubusercontent.com/19836026/96561619-162cfd00-12c0-11eb-8bd1-7054e75b373e.png)

Code to reproduce the plots:
```python
version = 'New'
t = np.arange(0, 20) * 1e9
y = np.sin(t * 1e-9)

s = Slice(TimeSeries(y, t))
ds = s.downsampled_to(.75)
ds2 = s.downsampled_to(.66)

plt.figure(figsize=(10, 7))
plt.subplot(2, 1, 1)
plt.title(f'{version} implementation')
plt.plot(s.timestamps, s.data)
plt.plot(ds.timestamps, ds.data, '.', markersize=12)
plt.plot(ds2.timestamps, ds2.data, '.', markersize=12)

plt.subplot(2, 1, 2)
plt.title('Residual between interpolated high frequency and downsampled')
plt.plot(ds.timestamps, ds.data - np.interp(ds.timestamps, s.timestamps, s.data))
plt.plot(ds2.timestamps, ds2.data - np.interp(ds2.timestamps, s.timestamps, s.data))

plt.savefig(f'{version}_ds.png')
```

For more samples per new sample, we can't use sines as test signals anymore, because the effect of the boxcar filtering starts suppressing the gain. But if we just use a single slope, we get:

![Old_ds](https://user-images.githubusercontent.com/19836026/96567049-9eae9c00-12c6-11eb-97c2-85a0a6ccf632.png)

![New_ds](https://user-images.githubusercontent.com/19836026/96567058-a0785f80-12c6-11eb-9517-bbb1eecee11d.png)

I do have worries about this whole non-integer resampling business though, which is:
1. What non-integer resampling rates do to the noise for low sample per point counts (and thereby to statistical models that may assume a single standard deviation for all points), since a different number of samples go into each sample (noise scaling with `1/sqrt(n)`). 
2. People using `.data` without realizing the timestamps are not equidistant. There should probably be a big warning about this in the docs. I would also consider emitting a warning for non-integer ratios in `downsampled_to`.
3. `np.sum` makes little sense when you don't guarantee the same sample count. It raises the question whether this setup with the `reduce` function makes much sense in this specific function.
4. Whether we should be using better downsampling methods that have better signal preservation below nyquist and better rejection above nyquist (brickwall rather than a simple boxcar filter). A higher quality resampling filter would be nice, but while this is pretty easy for `Continuous` channels, it can get very hairy for non-uniformly sampled `TimeSeries`.

But this worry applies to both methods for setting the timestamps.